File: majorProject/src/main/java/com/aocc/majorproject/Player.java
package com.aocc.majorproject;

import com.aocc.framework.Graphics;
import com.aocc.framework.PersonalMethods;
import com.aocc.framework.implementation.RotationHandler;

import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.Log;

public class Player {
	
	final float BORDER_WIDTH = 3;
	
	// Gyro Values
	final float SENSITIVITY = 100;		// tilt sensitivity
	final int MAX_SPEED = 30;
	final int MAX_RADIUS = 50;
	final int MIN_RADIUS = 10;
	
	// max/min values for health/shield
	final int MAX_HEALTH = 5;
	final int MAX_SHIELD = 100;
	final int MAX_OVERHEAT = 75;
	
	private boolean firstOverheat = true;
	private int tiltMode = 2;

	private float xBias = 0;
	private float yBias = -0.3f;
	private int health = 5;
	private int combo = 0;
	private int shield = MAX_SHIELD;
	
	private float characterDiameter = health * 15f;
	private float shieldWidth = shield/4;
	
	private float defaultX = 600;
	private float defaultY = 320;
	
	private float centerX = 0;
	private float centerY = 0;
	
	private float velocityX = 0;
	private float velocityY = 0;
	
	private float facingAngle = -90;
	
	private RectF mainCharacter = new RectF(0,0,0,0);
	private int overheat = 0;
	
	
	public void update() {
		// Rotation limited within 90 degree range, then turn into a decimal between 0 and 1 for easier modification.
		// Sensitivity determines the amount of tilting required for maximum speed. Bias determines default position.
		velocityX = (PersonalMethods.limitInside(RotationHandler.getRotationX(),-90,90)/90 + xBias)*SENSITIVITY;
		velocityX = PersonalMethods.limitInside(velocityX, -MAX_SPEED, MAX_SPEED);
		velocityY = (PersonalMethods.limitInside(RotationHandler.getRotationY(),-90,90)/90 + yBias)*SENSITIVITY;
		velocityY = PersonalMethods.limitInside(velocityY, -MAX_SPEED, MAX_SPEED);
		
		// angle character is facing: set now so it always faces the tilt direction
		if (velocityX >0) {
			facingAngle = (float) Math.toDegrees(Math.atan((velocityY/velocityX)));
		} else {
			facingAngle = 180 + (float) Math.toDegrees(Math.atan((velocityY/velocityX)));
		}
		
		//handling collisions with sides
		if ((defaultX + velocityX ) < BORDER_WIDTH){
			velocityX = 0;
			defaultX = BORDER_WIDTH;
		}
		
		if ((defaultY + velocityY) < BORDER_WIDTH){
			velocityY = 0;
			defaultY = BORDER_WIDTH;
		}
		
		if ((defaultX + velocityX) > 1280 - characterDiameter - BORDER_WIDTH){
			velocityX = 0;
			defaultX = 1280 - characterDiameter - BORDER_WIDTH;
		}
		
		if ((defaultY + velocityY ) > 720 - characterDiameter - BORDER_WIDTH){
			velocityY = 0;
			defaultY = 720 - characterDiameter - BORDER_WIDTH;
		}
		
//		//logs
//		if (GameScreen.getUpdateCount()%20 == 0){
//			Log.d("MainCharacter", "RotationX: " + RotationHandler.getRotationX());
//			Log.d("MainCharacter", "RotationY: " + RotationHandler.getRotationY());
//			Log.d("MainCharacter", "RotationZ: " + RotationHandler.getRotationZ());
//		}
		
		if (health <= 0){
			GameScreen.setGameOverFlag(true);
		}
		
		//increases overheat when shields aren't full (i.e. when off the powerup)
		if (shield < MAX_SHIELD){
			overheat = overheat - 1;
		}
		
		shield --;
		
		
		shieldWidth = shield/2;
		
		// sets the final values for the update
		health = (int)PersonalMethods.limitInside((float)health, 0, MAX_HEALTH);
		shield = (int)PersonalMethods.limitInside((float)shield, 0, MAX_SHIELD);
		overheat = (int)PersonalMethods.limitInside((float)overheat, 0, MAX_OVERHEAT);
		characterDiameter = MAX_RADIUS * health/5 + MIN_RADIUS;
		
		if (overheat == MAX_OVERHEAT){
			combo = 0;
			
			// plays sound only once
			if (firstOverheat == true){
				Assets.burn.play(MainMenuScreen.tapVol);
				firstOverheat = false;
			}
		} else {
			// resets for next overheat sound
			firstOverheat = true;
		}
		
		defaultX = defaultX + velocityX;
		defaultY = defaultY + velocityY;
		
		mainCharacter.set((int)defaultX, (int)defaultY, (int)defaultX + characterDiameter,
				(int)defaultY + characterDiameter);
		
		//to get a 'lag' effect, this can be placed before the previous statement for a 1-frame delay
		centerX = mainCharacter.centerX();		
		centerY = mainCharacter.centerY();
		
	}

	public void paint(Graphics g, Paint paint){
		paint.setStyle(Paint.Style.FILL);
		g.drawCircle(centerX, centerY, characterDiameter/2 + BORDER_WIDTH, Color.WHITE);
        g.drawCircle(centerX, centerY, characterDiameter/2 + shieldWidth, Color.argb(150, 0, 0, 255));
        g.drawArc(mainCharacter, facingAngle, 120, true, Color.BLUE);
		g.drawArc(mainCharacter, facingAngle + 120, 120, true, Color.RED);
		g.drawArc(mainCharacter, facingAngle + 240, 120, true, Color.GREEN);
        g.drawCircle(centerX, centerY, (int)characterDiameter/10, Color.WHITE);
        if (overheat >= MAX_OVERHEAT){
        	g.drawCircle(centerX, centerY, (int)characterDiameter/2 + BORDER_WIDTH + shieldWidth + 3, Color.argb(250, 255, 127, 39));
        	g.drawString("Overheating!", 100, 50, Color.MAGENTA, paint);
        } else {
        	g.drawCircle(centerX, centerY, (int)characterDiameter/2 + BORDER_WIDTH + shieldWidth + 3, Color.argb(150*overheat/MAX_OVERHEAT, 255, 127, 39));
        }
        
        // UI
        g.drawString("x" + combo, (int)mainCharacter.left, (int)mainCharacter.top, Color.WHITE, paint);
        
	}
	

	public int getCombo() {
		return combo;
	}

	public void setCombo(int combo) {
		this.combo = combo;
	}

	public int getOverheat() {
		return overheat;
	}

	public void setOverheat(int overheat) {
		this.overheat = overheat;
	}
	
	public float getCharacterDiameter() {
		return characterDiameter;
	}

	public void setCharacterDiameter(float characterDiameter) {
		this.characterDiameter = characterDiameter;
	}

	public float getShieldWidth() {
		return shieldWidth;
	}

	public void setShieldWidth(float shieldWidth) {
		this.shieldWidth = shieldWidth;
	}

	public int getShield() {
		return shield;
	}

	public void setShield(int shield) {
		this.shield = shield;
	}

	public float getShieldRadius() {
		return shieldWidth;
	}

	public void setShieldRadius(float shieldRadius) {
		this.shieldWidth = shieldRadius;
	}

	public float getDefaultX() {
		return defaultX;
	}


	public float getDefaultY() {
		return defaultY;
	}


	public float getVelocityX() {
		return velocityX;
	}


	public float getVelocityY() {
		return velocityY;
	}


	public void setDefaultX(int defaultX) {
		this.defaultX = defaultX;
	}


	public void setDefaultY(int defaultY) {
		this.defaultY = defaultY;
	}


	public void setVelocityX(float velocityX) {
		this.velocityX = velocityX;
	}


	public void setVelocityY(float velocityY) {
		this.velocityY = velocityY;
	}


	public float getCharacterRadius() {
		return characterDiameter;
	}

	public void setCharacterRadius(int characterRadius) {
		this.characterDiameter = characterRadius;
	}


	public void setDefaultX(float defaultX) {
		this.defaultX = defaultX;
	}

	public void setDefaultY(float defaultY) {
		this.defaultY = defaultY;
	}


	public RectF getMainCharacter() {
		return mainCharacter;
	}


	public void setMainCharacter(RectF mainCharacter) {
		this.mainCharacter = mainCharacter;
	}


	public float getFacingAngle() {
		return facingAngle;
	}


	public void setFacingAngle(float facingAngle) {
		this.facingAngle = facingAngle;
	}


	public float getCenterX() {
		return centerX;
	}


	public float getCenterY() {
		return centerY;
	}


	public void setCenterX(float centerX) {
		this.centerX = centerX;
	}


	public void setCenterY(float centerY) {
		this.centerY = centerY;
	}


	public float getBORDER_WIDTH() {
		return BORDER_WIDTH;
	}

	public int getHealth() {
		return health;
	}

	public void setHealth(int health) {
		this.health = health;
	}

	public void setCharacterRadius(float characterRadius) {
		this.characterDiameter = characterRadius;
	}	
	
	public float getxBias() {
		return xBias;
	}

	public void setxBias(float xBias) {
		this.xBias = xBias;
	}
	
	public float getyBias() {
		return yBias;
	}

	public void setyBias(float yBias) {
		this.yBias = yBias;
	}

	public int getTiltMode() {
		return tiltMode;
	}

	public void setTiltMode(int tiltMode) {
		this.tiltMode = tiltMode;
	}
	
}

File: majorProject/src/main/kotlin/com/aocc/majorproject/Player.kt
package com.aocc.majorproject

import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import com.aocc.framework.GameConstants
import com.aocc.framework.Graphics
import com.aocc.framework.PersonalMethods
import com.aocc.framework.implementation.RotationHandler
import kotlin.math.PI
import kotlin.math.atan

class Player(private val session: GameSession) {

    val borderWidth = 3f
    private val sensitivity = 100f
    private val maxSpeed = 30
    private val maxRadius = 50
    private val minRadius = 10

    private val maxHealth = 5
    private val maxShield = 100
    private val maxOverheat = 75

    private var firstOverheat = true
    var tiltMode = 2

    var xBias = 0f
    var yBias = -0.3f
    private var health = 5
    private var combo = 0
    private var shield = maxShield

    private var characterDiameter = health * 15f
    private var shieldWidth = shield / 4f

    private var defaultX = 600f
    private var defaultY = 320f

    private var centerX = 0f
    private var centerY = 0f

    private var velocityX = 0f
    private var velocityY = 0f

    private var facingAngle = -90f

    private var mainCharacter = RectF(0f, 0f, 0f, 0f)
    private var overheat = 0
    private var shieldDrainAccumulator = 0f

    /** Remaining invincibility (damage i-frames + flash) while > 0. Movement is unaffected. */
    private var iframeSeconds = 0f

    /** Debug-popup "god mode" toggle: damage is ignored entirely while true. */
    var debugInvincible = false

    fun update(deltaSeconds: Float) {
        val step = GameConstants.secondsToSteps(deltaSeconds)

        if (iframeSeconds > 0f) {
            iframeSeconds = maxOf(0f, iframeSeconds - deltaSeconds)
        }

        velocityX = (PersonalMethods.limitInside(RotationHandler.getRotationX(), -90, 90) / 90f + xBias) * sensitivity
        velocityX = PersonalMethods.limitInside(velocityX, -maxSpeed, maxSpeed)
        velocityY = (PersonalMethods.limitInside(RotationHandler.getRotationY(), -90, 90) / 90f + yBias) * sensitivity
        velocityY = PersonalMethods.limitInside(velocityY, -maxSpeed, maxSpeed)

        facingAngle = if (velocityX > 0) {
            (atan((velocityY / velocityX).toDouble()) * 180 / PI).toFloat()
        } else {
            180f + (atan((velocityY / velocityX).toDouble()) * 180 / PI).toFloat()
        }

        if (defaultX + velocityX * step < borderWidth) {
            velocityX = 0f
            defaultX = borderWidth
        }

        if (defaultY + velocityY * step < GameConstants.PLAY_AREA_TOP) {
            velocityY = 0f
            defaultY = GameConstants.PLAY_AREA_TOP.toFloat()
        }

        if (defaultX + velocityX * step > GameConstants.WORLD_WIDTH - characterDiameter - borderWidth) {
            velocityX = 0f
            defaultX = GameConstants.WORLD_WIDTH - characterDiameter - borderWidth
        }

        if (defaultY + velocityY * step > GameConstants.WORLD_HEIGHT - characterDiameter - borderWidth) {
            velocityY = 0f
            defaultY = GameConstants.WORLD_HEIGHT - characterDiameter - borderWidth
        }

        if (health <= 0) {
            session.setGameOverFlag(true)
        }

        if (shield < maxShield) {
            overheat = (overheat - step).toInt()
        }

        shieldDrainAccumulator += step
        while (shieldDrainAccumulator >= 1f) {
            shield--
            shieldDrainAccumulator -= 1f
        }

        shieldWidth = shield / 2f

        health = PersonalMethods.limitInside(health.toFloat(), 0, maxHealth).toInt()
        shield = PersonalMethods.limitInside(shield.toFloat(), 0, maxShield).toInt()
        overheat = PersonalMethods.limitInside(overheat.toFloat(), 0, maxOverheat).toInt()
        characterDiameter = maxRadius * health / 5f + minRadius

        if (overheat == maxOverheat) {
            combo = 0
            if (firstOverheat) {
                Assets.burn?.play(GamePreferences.getTapVolume().toFloat())
                firstOverheat = false
            }
        } else {
            firstOverheat = true
        }

        defaultX += velocityX * step
        defaultY += velocityY * step

        mainCharacter.set(
            defaultX.toInt().toFloat(),
            defaultY.toInt().toFloat(),
            defaultX.toInt() + characterDiameter,
            defaultY.toInt() + characterDiameter
        )

        centerX = mainCharacter.centerX()
        centerY = mainCharacter.centerY()
    }

    fun paint(g: Graphics, paint: Paint) {
        paint.style = Paint.Style.FILL
        g.drawCircle(centerX, centerY, characterDiameter / 2 + borderWidth, Color.WHITE)
        g.drawCircle(centerX, centerY, characterDiameter / 2 + shieldWidth, Color.argb(150, 0, 0, 255))
        g.drawArc(mainCharacter, facingAngle, 120f, true, Color.BLUE)
        g.drawArc(mainCharacter, facingAngle + 120f, 120f, true, Color.RED)
        g.drawArc(mainCharacter, facingAngle + 240f, 120f, true, Color.GREEN)
        g.drawCircle(centerX, centerY, (characterDiameter / 10), Color.WHITE)
        if (overheat >= maxOverheat) {
            g.drawCircle(
                centerX, centerY,
                characterDiameter / 2 + borderWidth + shieldWidth + 3,
                Color.argb(250, 255, 127, 39)
            )
            g.drawString("Overheating!", 100, 50, Color.MAGENTA, paint)
        } else {
            g.drawCircle(
                centerX, centerY,
                characterDiameter / 2 + borderWidth + shieldWidth + 3,
                Color.argb(150 * overheat / maxOverheat, 255, 127, 39)
            )
        }

        if (iframeSeconds > 0f) {
            // Pulsing red flash while invincible.
            val pulse = (iframeSeconds / IFRAME_SECONDS).coerceIn(0f, 1f)
            g.drawCircle(
                centerX, centerY,
                characterDiameter / 2 + borderWidth + shieldWidth + 8,
                Color.argb((200 * pulse).toInt().coerceIn(0, 255), 255, 60, 60)
            )
        }
    }

    /**
     * Applies a damaging hit: loses one health, resets combo and grants invincibility frames.
     * No-op while already invincible so a single collision can't chain-drain health.
     */
    fun onDamaged() {
        if (debugInvincible || iframeSeconds > 0f) {
            return
        }
        health -= 1
        combo = 0
        iframeSeconds = IFRAME_SECONDS
    }

    fun isInvincible(): Boolean = iframeSeconds > 0f

    fun getCombo(): Int = combo

    fun setCombo(combo: Int) {
        this.combo = combo
    }

    fun getOverheat(): Int = overheat

    fun setOverheat(overheat: Int) {
        this.overheat = overheat
    }

    fun getCharacterDiameter(): Float = characterDiameter

    fun setCharacterDiameter(characterDiameter: Float) {
        this.characterDiameter = characterDiameter
    }

    fun getShieldWidth(): Float = shieldWidth

    fun setShieldWidth(shieldWidth: Float) {
        this.shieldWidth = shieldWidth
    }

    fun getShield(): Int = shield

    fun setShield(shield: Int) {
        this.shield = shield
    }

    fun getShieldRadius(): Float = shieldWidth

    fun setShieldRadius(shieldRadius: Float) {
        shieldWidth = shieldRadius
    }

    fun getDefaultX(): Float = defaultX

    fun getDefaultY(): Float = defaultY

    fun getVelocityX(): Float = velocityX

    fun getVelocityY(): Float = velocityY

    fun setDefaultX(defaultX: Int) {
        this.defaultX = defaultX.toFloat()
    }

    fun setDefaultY(defaultY: Int) {
        this.defaultY = defaultY.toFloat()
    }

    fun setVelocityX(velocityX: Float) {
        this.velocityX = velocityX
    }

    fun setVelocityY(velocityY: Float) {
        this.velocityY = velocityY
    }

    fun getCharacterRadius(): Float = characterDiameter

    fun setCharacterRadius(characterRadius: Int) {
        characterDiameter = characterRadius.toFloat()
    }

    fun setDefaultX(defaultX: Float) {
        this.defaultX = defaultX
    }

    fun setDefaultY(defaultY: Float) {
        this.defaultY = defaultY
    }

    fun getMainCharacter(): RectF = mainCharacter

    fun setMainCharacter(mainCharacter: RectF) {
        this.mainCharacter = mainCharacter
    }

    fun getFacingAngle(): Float = facingAngle

    fun setFacingAngle(facingAngle: Float) {
        this.facingAngle = facingAngle
    }

    fun getCenterX(): Float = centerX

    fun getCenterY(): Float = centerY

    fun setCenterX(centerX: Float) {
        this.centerX = centerX
    }

    fun setCenterY(centerY: Float) {
        this.centerY = centerY
    }

    fun getBORDER_WIDTH(): Float = borderWidth

    fun getHealth(): Int = health

    fun setHealth(health: Int) {
        this.health = health
    }

    fun setCharacterRadius(characterRadius: Float) {
        characterDiameter = characterRadius
    }

    companion object {
        /** Invincibility / flash window applied on taking damage. */
        const val IFRAME_SECONDS = 0.5f
    }
}
